Mathlib/NumberTheory/Padics/MahlerBasis.lean

```diff
@@ -0,0 +1,73 @@
+/-
+Copyright (c) 2024 David Loeffler. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Giulio Caflisch, David Loeffler
+-/
+import Mathlib.Analysis.Normed.Group.Ultra
+import Mathlib.Data.Nat.Factorial.BigOperators
+import Mathlib.NumberTheory.Padics.ProperSpace
+import Mathlib.Topology.ContinuousMap.Compact
+
+/-!
+# The Mahler basis of continuous functions
+
+In this file we introduce the Mahler basis function `mahler k`, for `k : ℕ`, which is the unique
+continuous map `ℤ_[p] → ℚ_[p]` agreeing with `n ↦ n.choose k` for `n ∈ ℕ`.
+
+In future PR's, we will show that these functions give a Banach basis for the space of continuous
+maps `ℤ_[p] → ℚ_[p]`.
+
+## References
+
+* [P. Colmez, *Fonctions d'une variable p-adique*][colmez2010]
+-/
+
+open Finset
+
+variable {p : ℕ} [hp : Fact p.Prime]
+
+/--
+The `k`-th Mahler basis function, i.e. the unique continuous function `ℤ_[p] → ℚ_[p]`
+agreeing with `n ↦ n.choose k` for `n ∈ ℕ`. See [colmez2010], §1.2.1.
+-/
+noncomputable def mahler (k : ℕ) : C(ℤ_[p], ℚ_[p]) where
+  toFun x := (∏ i in range k, (x.1 - i)) / k.factorial
```

Comment:
The underlying function is `Ring.choose x k` (from `RingTheory.Binomial`). I'm not sure if you want to go through the trouble of producing a `BinomialRing` instance for `Z_[p]` and proving continuity with this alternative definition, though.  It would however make `mathler_natCast_eq` superfluous.

Reply:
We wondered about using the `BinomialRing` machinery but didn't pursue it, since I wasn't sure how easy it would be to prove continuity from this definition. But if having a BinomialRing instance on Zp is something that would be useful in its own right then we can try that route instead. (I think there are no other examples of BinomialRings at present except N and Z, so having another one can't do any harm.)

Reply:
So I tried, and it wasn't much fun. Obtaining a `BinomialRing` instance on Zp is not too bad, it's just a reorganisation of the code we already have for showing that `‖mahler k‖ = 1`. But if you define `mahler k` using `Ring.choose`, deriving continuity from this seems to be quite awkward; I got as far as proving a goal which printed as `(descPochhammer ℤ k).smeval x = (descPochhammer ℤ k).smeval x`, for `x : ℤ_[p]`, but unfortunately the LHS and RHS seem to be using two different definitions of `smul` which Mathlib is unable to show are the same.

Reply:
See #18410.